### ISSUE TITLE
Concat not checking for broadcast

### ISSUE THREAD

**Issue by Atticus1806** [rwth-i6/returnn_common#161]
I noticed that returnn_common does not check dimensions when concating during network construction. 
So when doing: `concat((a, a.feat), (b, b.feat))` with `a.shape = [B, T, F]` and `b.shape = [B,F]` this does not crash in network construction but only when running the network. Of course setting `allow_broadcast=True` fixes it. 
This issue is because I am wondering whether this is intended to not be checked in returnn_common (like e.g. it is done for ops like `+`). 

Below is an example test. The linear layer and sparse dims are just so that `dummy_run_net` works without modifications, of course in a real test this could be changed. 

```python
def test_concat_broadcasting():

  class _Net(nn.Module):

    def __init__(self, feat: nn.Dim):
      super(_Net, self).__init__()
      self.linear = nn.Linear(out_dim=feat)

    def __call__(self, a: nn.Tensor, b: nn.Tensor) -> nn.Tensor:
      a = self.linear(a)
      b = self.linear(b)
      return nn.concat((a, a.feature_dim), (b, b.feature_dim))  # This crashes in RETURNN

  nn.reset_default_root_name_ctx()
  feat = nn.FeatureDim('F', 3)
  time = nn.SpatialDim("time")
  with_time = nn.get_extern_data(nn.Data('data', dim_tags=[nn.batch_dim, time], sparse_dim=feat))
  no_time = nn.get_extern_data(nn.Data('classes', dim_tags=[nn.batch_dim], sparse_dim=feat))
  net = _Net(feat)
  out = net(with_time, no_time)
  out.mark_as_default_output()
  config_code = nn.get_returnn_config().get_complete_py_code_str(net)
  from returnn_common.tests.returnn_helpers import config_net_dict_via_serialized, dummy_run_net
  config, net_dict = config_net_dict_via_serialized(config_code)
  dummy_run_net(config)

**Comment by albertz** [comment]
On early error checks on returnn-common side:

- Yes, whenever this is easy and reasonable to do, we should do it.
- It's not always easy, and esp not for the auto-generated layer wrappers.
- In this case, I think we can do it, as `concat` is explicitly defined, not automatically wrapped, and such check should not be too complicated.

For allowing broadcasting, we can also differentiate two cases:

- At least one source has all dims: [B,F] + [B,T,F] -> [B,T,F]
- Broadcasting for all sources: [B,T1,F] + [B,T2,F] -> [B,T1,T2,F]

`ConcatLayer` on RETURNN is a bit inconsistent to `CombineLayer`/`CompareLayer`:

- `CombineLayer`/`CompareLayer`: Allows broadcasting automatically when there is at least one source with all dims. Only requires the flag `allow_broadcast_all_sources` when broadcasting is needed for all sources. This is done this way because the first case is usually always safe but the second case might be a potential error by the user.
- `ConcatLayer`: Never allows any broadcasting automatically. The flag `allow_broadcast` allows then both kinds of broadcasting. This is done this way because broadcasting is kind of unexpected, and also more expensive here (not natively supported by TF so we first need to expand dims and tile the missing dims). I'm not sure though whether we maybe better should have differentiated between those two cases of broadcasting. We can still add that, via a new behavior version. This layer is probably anyway not used much yet.
